packages/react-dom-bindings/src/server/ReactFlightServerConfigDOM.js

```diff
@@ -7,6 +7,110 @@
  * @flow
  */
 
+import {enableFloat} from 'shared/ReactFeatureFlags';
+
+import {dispatchDirective} from 'react-server/src/ReactFlightDirectives';
+
+import ReactDOMSharedInternals from 'shared/ReactDOMSharedInternals';
+const ReactDOMCurrentDispatcher = ReactDOMSharedInternals.Dispatcher;
+
+const ReactDOMFlightServerDispatcher = {
+  prefetchDNS,
+  preconnect,
+  preload,
+  preinit,
+};
+
+export function prepareHostDispatcher(): void {
+  ReactDOMCurrentDispatcher.current = ReactDOMFlightServerDispatcher;
+}
+
 // Used to distinguish these contexts from ones used in other renderers.
 // E.g. this can be used to distinguish legacy renderers from this modern one.
 export const isPrimaryRenderer = true;
+
+type PrefetchDNSDirective = ['prefetchDNS', string];
+type PreconnectDirective =
+  | ['preconnect', string]
+  | ['preconnect', string, {crossOrigin?: string}];
+type PreloadDirective = ['preload', string, PreloadOptions];
+type PreinitDirective = ['preinit', string, PreinitOptions];
+
+export type Directive =
+  | PrefetchDNSDirective
+  | PreconnectDirective
+  | PreloadDirective
+  | PreinitDirective;
+
+let didWarnAsyncEnvironmentDev = false;
+
+function dispatch(directive: Directive) {
+  const successful = dispatchDirective(directive);
+  if (__DEV__) {
+    if (!successful && !didWarnAsyncEnvironmentDev) {
+      didWarnAsyncEnvironmentDev = true;
+      const methodName = `ReactDOM.${directive[0]}()`;
+      console.error(
+        '%s: React expected to be able to associate this call to a specific Request but cannot. It is possible that this call was invoked outside of a React component. If you are calling it from within a React component that is an async function after the first `await` then you are in an environment which does not support AsyncLocalStorage. In this kind of environment %s does not do anything when called in an async manner. Try moving this function call above the first `await` within the component or remove this call. In environments that support AsyncLocalStorage such as Node.js you can call this method anywhere in a React component even after `await` operator.',
```

Comment:
It should be a noop to call it in module init, even if that module init is lazy and happens later. This would error in that case which doesn't seem right.

We already silently fail to `cache()` if you're outside `AsyncLocalStorage` for this reason. Since both of these are optimistic, it seems fine.